submariner/globalnet-enhancement2-0.md

```diff
@@ -0,0 +1,197 @@
+
+Similarly, a ClusterIP Service is annotated with a globalIp so that remote Pods can communicate
+with it using its globalIp. Globalnet programs the necessary ingress rules on the active Gateway
+node of the cluster, where the service resides. The ingress rules enable any incoming traffic
+destined to the globalIp of the Service to be redirected to the corresponding ClusterIP Service.
+
+In order to support connectivity from the Host Network to the remote services, Globalnet also
+annotates every node in the cluster with a unique globalIp.
+
+While this works fine for small clusters, where the number of Pods and Services are limited, for
+large clusters it poses scalability and performance issues as we would eventually run out of
+globalIps because of annotating every Pod/Service/Node object. Also, in this design, Pod to Pod
+connectivity as well as Headless Services are not supported.
+
+## Proposal
+
+This enhancement proposes the following changes to make Globalnet implementation scale better.
+
+1. By default, every cluster will be assigned a single globalIp which will be used as egress IP
+   for any cross-cluster communication.
+2. Provision to allocate a globalIp at the namespace/project level which takes precedence over
+   the globalIp allocated to the cluster.
+3. Provision to allocate a globalIp to a Pod (or a set of Pods) which has higher precedence
+   over globalIPs at namespace level. If there are multiple policies, users can specify a priority
+   field for each of the policy.
+4. Instead of annotating every ClusterIP Service in the cluster, only exported Services will be
+   annotated with a globalIp.
+5. For Headless Services, necessary ingress rules will be programmed to support connectivity to
+   the backend Pods. In a Headless Service use-case, every backend Pod will be annotated with
+   a unique globalIP and the same globalIP will be used for both Ingress as-well-as Egress.
+   globalIPs allocated for Pods backed by Headless Services will have highest priority over other
+   egress policies.
+
+![GlobalnetTopology](./images/globalnet-topology.png)
+
+If the external application in a remote Cluster is interested in consuming the exported
+Service, it will have to reach the globalIP assigned to the exportedService and the response
+from the backend Pods will have the source as the globalIP of the exportedService.
+
+On the other hand, when the backend Pod tries to talk to the external application in remote
+Cluster, the external application will see that traffic is coming with a globalIP that is
+assigned either at the Cluster level, or namespace level, or the one requested by the user
+at Pod level. This asynchronous behavior will be analogous to the behavior seen with standard
+K8s applications when they talk to any Services. Applications desiring to have the same IPaddress
+for both ingress and egress traffic can create Headless Services in front of them.
+
+The proposal has its own set of Pros and Cons.
```

Comment:
Currently, while deploying Submariner Globalnet, we can provide an env variable (SUBMARINER_EXCLUDE_NS) to inform Globalnet to Skip annotating any Pods/Services in that Namespace. 

In the new design, do we want to continue to support it or deprecate it given that we are now using CRDs to annotate at Namespace/Pod level. Any suggestions?

Reply:
Related: Based on the discussion with @mangelajo, to support Globalnet2.0 we will have to break the backward compatibility as we might be using two annotations (one for Egress and the other for Ingress).

Reply:
We can deprecate that. I think it could be useful to let users mark specific system namespaces sometimes, I see use cases for that. (like connecting to a remote API, etc..)

Reply:
Thanks @mangelajo. So, we'll deprecate SUBMARINER_EXCLUDE_NS.

Regarding, "I think it could be useful to let users mark specific system namespaces sometimes" -> For this, are you suggesting that we support a new Env variable that instructs Globalnet to automatically annotate all Services in those namespaces? If so, in most of the common use-cases we may not really require accessing those Services and even if there is a need, user can explicitly export those services right? WDYT?

Reply:
+1 to deprecate `SUBMARINER_EXCLUDE_NS`.

> I think it could be useful to let users mark specific system namespaces sometimes

Should not be an ENV variable. This is something that was briefly discussed in MCS API. We can have an annotation or label added to a namespace, say `NamespaceExport`, and then we auto-export all services in that namespace. Export of service is what will lead to globalIP for it. This retains design approach of requiring to export services, but providing a sort of mechanism to export services in a namespace.

Reply:
Good point. I'd prefer that we do that as a future enhancement (should not be too difficult once the basic implementation is in place). Currently, AFAICT we will have a lot of work from an implementation POV to support what is specified in this enhancement.